app/assets/javascripts/govuk_publishing_components/lib/auto-scroll-tracker.js

```diff
@@ -0,0 +1,170 @@
+window.GOVUK = window.GOVUK || {}
+window.GOVUK.Modules = window.GOVUK.Modules || {};
+
+(function (Modules) {
+  function AutoScrollTracker ($module) {
+    this.$module = $module
+    this.SCROLL_TIMEOUT_DELAY = 20
+    this.RESIZE_TIMEOUT_DELAY = 100
+    this.PAGEHEIGHT_TIMEOUT_DELAY = 500
+    this.pageHeight = document.querySelector('body').clientHeight
+    this.trackedNodes = []
+    this.config = {
+      ignoreHeadingsInside: ['.govuk-cookie-banner', '.gem-c-feedback', 'footer'],
+      percentages: [20, 40, 60, 80, 100]
+    }
+  }
+
+  AutoScrollTracker.prototype.init = function () {
+    var consentCookie = window.GOVUK.getConsentCookie()
+
+    if (consentCookie && consentCookie.settings) {
+      this.startModule()
+    } else {
+      this.startModule = this.startModule.bind(this)
+      window.addEventListener('cookie-consent', this.startModule)
+    }
+  }
+
+  AutoScrollTracker.prototype.startModule = function () {
+    if (!window.GOVUK.analyticsVars.scrollTrackerStarted) {
+      window.GOVUK.analyticsVars.scrollTrackerStarted = true
+
+      this.trackType = this.$module.getAttribute('data-track-type')
+      if (this.trackType === 'headings') {
+        this.track = new AutoScrollTracker.Heading(this.config)
+      } else {
+        this.track = new AutoScrollTracker.Percentage(this.config)
+      }
+
+      this.getWindowDetails()
+
+      if (this.trackedNodes) {
+        // store event listener functions as variables so they can be removed if needed
+        this.scrollEvent = this.onScroll.bind(this)
+        window.addEventListener('scroll', this.scrollEvent)
+        this.resizeEvent = this.onResize.bind(this)
+        window.addEventListener('resize', this.resizeEvent)
+
+        // check if the page height changes e.g. accordion opened
+        this.interval = window.setInterval(function () {
+          var pageHeight = document.querySelector('body').clientHeight
+          if (pageHeight !== this.pageHeight) {
+            this.pageHeight = pageHeight
+            this.getWindowDetails()
+          }
+        }.bind(this), this.PAGEHEIGHT_TIMEOUT_DELAY)
+      }
+    }
+  }
+
+  AutoScrollTracker.prototype.onScroll = function onScroll () {
```

Comment:
Out of curiosity, what's the difference between
```js
AutoScrollTracker.prototype.onScroll = function onScroll () { ... }
```
and 
```js
AutoScrollTracker.prototype.onScroll = function () { ... }
```
?

Reply:
Ah, good spot. The difference is that the first one isn't anonymous. Or something like that. It's kinda weird. Even though both of them are called `onScroll` as far as referring to it inside the module is concerned, only the first example actually has a function name of `onScroll`.

I meant to take this out - it was part of getting the tests to work properly, but then I found a different way to remove the scroll listener after each test. I'll remove, thanks.

Reply:
Removed.